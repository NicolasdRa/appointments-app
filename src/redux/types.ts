export interface User {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  memberSince: string;
  dateOfBirth: string;
  insurance: string;
  img: string;
}

export interface Doctor {
  id: string;
  title?: string;
  firstName?: string;
  lastName?: string;
  gender?: string;
  email?: string;
  dateOfBirth?: string;
  registerDate?: string;
  phone?: string;
  picture?: string;
  location?: string;
}

export interface Appointment {
  id: string;
  title: string;
  date: string;
  time: string;
  user: string;
  doctor?: string;
  description?: string;
}
