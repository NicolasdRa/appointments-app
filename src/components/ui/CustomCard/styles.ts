import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      minWidth: 200,
      maxWidth: 210,
      display: 'flex',
      flexDirection: 'column',
      margin: '0.5rem',
      // minWidth: '14rem',
      overflow: 'visible',

      [theme.breakpoints.up('md')]: {
        maxWidth: '16rem',
      },
    },

    media: {
      height: 130,
      borderRadius: '6px',
      margin: '1rem',
    },

    btn: {
      color: theme.palette.common.beige,
    },
  }),
);
