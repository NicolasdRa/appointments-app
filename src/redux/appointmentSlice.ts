import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import type { RootState } from './store';
import { Appointment } from './types';
import { initialAppointments } from '../assets/data';

// SLICE

// type for the slice state
interface AppointmentsState {
  data: Appointment[];
}

// initial state
const initialState: AppointmentsState = { data: initialAppointments };

export const appointmentSlice = createSlice({
  name: 'appointments',
  initialState,
  reducers: {
    create: (state, action: PayloadAction<Appointment>) => {
      state.data.push(action.payload);
    },
    cancel: (state, action: PayloadAction<Appointment>) => {
      const filtered = state.data.filter(
        (item) => item.id !== action.payload.id,
      );
      state.data = filtered;
    },
  },
});

// ACTIONS
export const { create, cancel } = appointmentSlice.actions;

// SELECTORS
export const selectAppointments = (state: RootState) => state.appointments.data;

// TODO: create selector that returns appointments by user

export default appointmentSlice.reducer;
