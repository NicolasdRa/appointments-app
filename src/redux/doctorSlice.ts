import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import axios from 'axios';

import type { RootState } from './store';
import { Doctor } from './types';
import { APP_ID, baseURL } from '../constants';

// THUNKS
export const fetchDoctors = createAsyncThunk('doctors/fetch', async () => {
  const res = await axios({
    method: 'GET',
    url: `${baseURL}`,
    headers: {
      'app-id': `${APP_ID}`,
    },
    responseType: 'json',
  });

  return res.data.data as Doctor[];
});

// SLICE
export interface DoctorState {
  loading: 'idle' | 'pending' | 'succeeded' | 'failed';
  data: Doctor[] | [];
  current: Doctor | undefined;
}

const initialState: DoctorState = {
  loading: 'idle',
  data: [],
  current: undefined,
};

const doctorSlice = createSlice({
  name: 'doctors',
  initialState,
  reducers: {
    setCurrent: (state, action: PayloadAction<Doctor>) => {
      state.current = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchDoctors.pending, (state, action) => {
      state.loading = 'pending';
    });
    builder.addCase(fetchDoctors.fulfilled, (state, action) => {
      state.loading = 'succeeded';
      state.data = action.payload;
    });
    builder.addCase(fetchDoctors.rejected, (state, action) => {
      state.loading = 'failed';
    });
  },
});

// ACTIONS
export const { setCurrent } = doctorSlice.actions;

// SELECTORS
export const selectDoctors = (state: RootState) => state.doctors.data;

export const selectCurrentDoctor = (state: RootState) => state.doctors.current;

export default doctorSlice.reducer;
