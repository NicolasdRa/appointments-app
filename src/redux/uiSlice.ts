import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import type { RootState } from './store';

// SLICE
interface UiState {
  formOpen: boolean;
}

const initialState: UiState = {
  formOpen: false,
};

const uiSlice = createSlice({
  name: 'ui',
  initialState,
  reducers: {
    toogleFormOpen: (state, action: PayloadAction) => {
      const open = state.formOpen;
      state.formOpen = !open;
    },
  },
});

// ACTIONS
export const { toogleFormOpen } = uiSlice.actions;

// SELECTORS
export const selectFormOpen = (state: RootState) => state.ui.formOpen;

export default uiSlice.reducer;
