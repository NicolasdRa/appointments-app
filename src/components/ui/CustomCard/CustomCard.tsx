import { Link } from 'react-router-dom';

import {
  Button,
  Card,
  CardActionArea,
  CardActions,
  CardContent,
  CardMedia,
  Typography,
} from '@material-ui/core';

import { useStyles } from './styles';

import { Doctor } from '../../../redux/types';
import { useAppDispatch } from '../../../hooks/hooks';
import { setCurrent } from '../../../redux/doctorSlice';

interface CustomCardProps {
  key: string;
  doctor: Doctor;
}

export const CustomCard = ({ key, doctor }: CustomCardProps) => {
  const classes = useStyles();
  const dispatch = useAppDispatch();

  const { title, firstName, lastName, picture } = doctor;

  const handleClickSetCurrentDoctor = (): void => {
    dispatch(setCurrent(doctor));
  };

  return (
    <Card className={classes.root}>
      <CardActionArea>
        <CardMedia className={classes.media} image={picture} title='Doctor' />
        <CardContent>
          <Typography gutterBottom variant='subtitle1' component='h2'>
            {`${title} ${firstName} ${lastName}`}
          </Typography>
          <Typography variant='body2' color='textSecondary' component='p'>
            Berlin - Kreuzberg
          </Typography>
          <Typography variant='body2' color='textSecondary' component='p'>
            general practitioner
          </Typography>
        </CardContent>
      </CardActionArea>
      <CardActions>
        <Button
          component={Link}
          to={`doctor/${lastName}`}
          onClick={handleClickSetCurrentDoctor}
          className={classes.btn}>
          Make an appoinment
        </Button>
      </CardActions>
    </Card>
  );
};
