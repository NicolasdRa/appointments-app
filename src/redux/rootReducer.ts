import { combineReducers } from 'redux';

import doctorSlice from './doctorSlice';
import appointmentSlice from './appointmentSlice';
import uiSlice from './uiSlice';

const rootReducer = combineReducers({
  doctors: doctorSlice,
  appointments: appointmentSlice,
  ui: uiSlice,
});

export default rootReducer;
