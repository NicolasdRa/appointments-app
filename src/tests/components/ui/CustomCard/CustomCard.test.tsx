import { CustomCard } from '../../../../components/ui/CustomCard/CustomCard';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import { store } from '../../../../redux/store';

describe('tests on <CustomCard/>', () => {
  const key = '1';
  const doctor = {
    id: '0F8JIqi4zwvb77FGz6Wt',
    lastName: 'Fiedler',
    firstName: 'Heinz-Georg',
    email: 'heinz-georg.fiedler@example.com',
    title: 'mr',
    picture: 'https://randomuser.me/api/portraits/men/81.jpg',
  };
  const wrapper = shallow(
    <Provider store={store}>
      <CustomCard key={key} doctor={doctor} />
    </Provider>,
  );

  test('should render <CustomCard/> correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
