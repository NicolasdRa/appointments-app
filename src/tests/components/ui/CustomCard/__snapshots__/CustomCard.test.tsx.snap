// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on <CustomCard/> should render <CustomCard/> correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <CustomCard
    doctor={
      Object {
        "email": "heinz-georg.fiedler@example.com",
        "firstName": "Heinz-Georg",
        "id": "0F8JIqi4zwvb77FGz6Wt",
        "lastName": "Fiedler",
        "picture": "https://randomuser.me/api/portraits/men/81.jpg",
        "title": "mr",
      }
    }
    key="1"
  />
</ContextProvider>
`;
